### ISSUE TITLE
Zenphoto 1.5 is failing to generate GIF thumbnails with Imagick turned on

### ISSUE THREAD

**Issue by JatB** [zenphoto/zenphoto#1189]
Here is what my debug log says (with the personal file path info edited out):

```
{298217:Tue, 07 Aug 2018 10:37:39 GMT} Zenphoto v1.5
  NOTICE: Undefined index: dataprivacy_policy_zenpage in /fullpath/zp-core/admin-options.php on line 475

{298329:Tue, 07 Aug 2018 10:37:45 GMT}
  cacheImage(/public_directory1/a4274ed0f2ad366efa9c3411c67cfa72cf1f7ed9.ImageA_200_thumb.gif) exception: memory allocation failed `/fullpath/cache/public_directory1/a4274ed0f2ad366efa9c3411c67cfa72cf1f7ed9.ImageA_200_thumb.gif' @ error/gif.c/WriteGIFImage/1609
{798648:Sun, 12 Aug 2018 12:17:01 GMT}
  cacheImage(/public_directory2/eb106bf3d77001aabfdcbf505662e2dc90437b66.ImageB_200_thumb.gif) exception: memory allocation failed `/fullpath/cache/public_directory2/eb106bf3d77001aabfdcbf505662e2dc90437b66.ImageB_200_thumb.gif' @ error/gif.c/WriteGIFImage/1609
{798953:Sun, 12 Aug 2018 12:17:06 GMT}
  cacheImage(/public_directory2/89c85c93e1265718423280b72f417f2b5c58c1d7.ImageC_200_thumb.gif) exception: memory allocation failed `/fullpath/cache/public_directory2/89c85c93e1265718423280b72f417f2b5c58c1d7.ImageC_200_thumb.gif' @ error/gif.c/WriteGIFImage/1609
```

This is what version of Imagick the server is running:
> Graphics support: PHP Imagick library 3.1.2
> ImageMagick 6.7.6-8 2016-07-19 Q16 http://www.imagemagick.org	
>     supporting: bmp, bmp2, bmp3, gif, gif87, jpeg, jpg, png, png24, png32, png8, tiff, tiff64

If I turn off Imagick and use GD, then clear the cache, the GIF thumbnails will be generated.

This is what version of GD the server is running:
> Graphics support: PHP GD library bundled (2.1.0 compatible)	
>     supporting: gif, jpg, jpeg, png, wbmp

This is not a typical memory issue, where the image is too large and can't render because of the size. These GIFs are really small, 320x400 for example, and everything else that is much larger is being generated. I also know that it worked in the past, but I don't know exactly what version it might have stopped working on. I only reset the cache with the latest 1.5 to clear out space for the new thumbnails.

I don't think I will be able to test the latest support build on this server, but I might be able to test it somewhere else at a later time.

**Comment by acrylian** [comment]
No idea what that could be. Can you provide the actual images causing the issue so we can try to reproduce it? 

Thanks for also including the surely unrelated index noticed of the dataprivacy_policy. I will take a look at that too;-)

It would be good if you can create a staging site somewhere. If we can reproduce and fix anything it would be great to know it is actually before releasing 1.5.1. We are working on some image/cacheManager related stuff currently anyway.

**Comment by fretzl** [comment]
I'm curious about your images too.
So far I have not been able to reproduce this on my local MAMP
```
Graphics support: PHP Imagick library 3.4.3
ImageMagick 6.9.6-2 Q16 x86_64 2017-07-03 http://www.imagemagick.org	
```

**Comment by JatB** [comment]
Long story short, it must be something with my hosting provider's setup. If I hear anything back from their tech support, I will let you know.


Long story long, what steps I tried to troubleshoot the issue…

I tested it in the latest MAMP v5.0.1, setting Apache to port 80.

I installed Zenphoto 1.4.14 first. Tried the GIFs with both GD and Imagick enabled, with the thumbnails cropped and not cropped. Both worked fine, with the difference being Imagick created animated GIF thumbnails and GD didn't.

Tried it with Zenphoto 1.5 and it worked.

I even tried turning on "Obscure cache filenames" in Options-Security to try to match my live setup.

MAMP 5.0.1 is running a newer version of Imagick, so I don't know if that is it.

> Graphics support: PHP Imagick library 3.4.3
> ImageMagick 6.9.6-2 Q16 x86_64 2018-07-17 http://www.imagemagick.org	
>     supporting: bmp, bmp2, bmp3, gif, gif87, jpeg, jpg, png, png24, png32, png8, tiff, tiff64

The GD is the same version of GD-

> Graphics support: PHP GD library bundled (2.1.0 compatible)	
>     supporting: gif, jpg, jpeg, png, wbmp

I then tried going back multiple older versions of MAMP to try to get close to the old version of Imagick that my cutting edge hosting company is running.

In MAMP 3.5.2-

> Graphics support: PHP Imagick library 3.1.2
> ImageMagick 6.8.9-7 Q16 x86_64 2015-07-06 http://www.imagemagick.org	
>     supporting: bmp, bmp2, bmp3, gif, gif87, jpeg, jpg, png, png24, png32, png8, tiff, tiff64 

> PHP version: 5.6.10
> MySQLi version: 5.5.42

With the same version of Imagick running (though a slightly newer version of ImageMagick) in MAMP 3.5.2, GIF thumbnail creation works in Zenphoto 1.4.14 and 1.5.

I then tried a whole new Zenphoto install on my live server; new database, new everything, starting with the previous version of Zenphoto 1.4.14. I even tried a different FTP client, just to be sure that wasn't the issue. It still did not work with Imagick enabled.

There was nothing special about the GIFs, some were static, some animated, they all failed.

I then spent two hours chatting with my hosting provider's tech support. Which turned up nothing, of course. It didn't help that they thought Zenphoto was something to do with Wordpress, even though they offer Zenphoto as part of their one-click install option. It is version 1.4.5.9 from over 4 years ago, and not what I'm using.

They kept saying the issue was your problem. Also, that if it was an actual issue with the PHP Imagick or ImageMagick they had installed, that it also wasn't their problem, even though they install and enable them by default.


Not really sure what you guys can do, or me, besides switching my hosting provider, which is a real hassle. I finally had them escalate the issue, not that it will probably ever come to anything, but if it does, I'll keep you posted.

**Comment by acrylian** [comment]
Thanks for the follow up. I am currently on MAMP 4.4.x still but I somehow doubt this is the issue. There has not been any significant change to the direct image handling at all…

> They kept saying the issue was your problem. Also, that if it was an actual issue with the PHP Imagick or ImageMagick they had installed, that it also wasn't their problem, even though they install and enable them by default.

I can sort of understand it if it is not a configuration issue on their end since they didn't make these tools. However, server configuration is apparently a quite complicated task…

It still could be the images somehow.Althoug gifs don't really have meta/exif data, it could be something else…

**Comment by JatB** [comment]
To try to troubleshoot the issue as much as I can, I also tried changing the versions of PHP to see if that would matter, or if they would use a different version of Imagick.

Starting in Zenphoto 1.4.14 of my test version on my live server (to see if anything was different), the PHP version was changed from 5.6.30 to 7.0.25 for the test version only and Imagick was enabled. I immediately got this error message when I applied the option in the Options-Image tab-

`Fatal error: Out of memory (allocated 4194304) (tried to allocate 8192 bytes) in /fullpath/zenphototest/zp-core/class-album.php on line 1670`

If I tried to go to the gallery itself, I got this message-

`Fatal error: Out of memory (allocated 4194304) (tried to allocate 32768 bytes) in /fullpath/zenphototest/zp-core/lib-auth.php on line 538`

Nothing new in the debug log, but almost the same thing in the PHP error log (with the entries repeated)-

```
[14-Aug-2018 09:46:37 America/Chicago] PHP Fatal error:  Out of memory (allocated 2883584) (tried to allocate 12288 bytes) in /fullpath/zenphototest/zp-core/class-album.php on line 1556
[14-Aug-2018 23:23:03 America/Chicago] PHP Fatal error:  Out of memory (allocated 4194304) (tried to allocate 32768 bytes) in /fullpath/zenphototest/zp-core/lib-auth.php on line 538
```
Memory shouldn't be an issue, as I'm not really using any-

![cpanel - usage](https://user-images.githubusercontent.com/10157440/44135150-b6fd744e-a01c-11e8-9b92-89187fef3a50.png)

If I turned off imagick in the php.ini, the site would load.

I then upgraded the site to 1.5 and tried again. This time blank pages, no error messages, even though display_errors is turned on in the php.ini.

In the PHP error log was this-

```
[15-Aug-2018 00:10:19 America/Chicago] PHP Fatal error:  Out of memory (allocated 4194304) (tried to allocate 32768 bytes) in /fullpath/zenphototest/zp-core/template-functions.php on line 1737
[15-Aug-2018 00:10:25 America/Chicago] PHP Fatal error:  Out of memory (allocated 4194304) (tried to allocate 32768 bytes) in /fullpath/zenphototest/zp-core/template-functions.php on line 1737
[15-Aug-2018 00:10:34 America/Chicago] PHP Fatal error:  Out of memory (allocated 4194304) (tried to allocate 32768 bytes) in /fullpath/zenphototest/zp-core/template-functions.php on line 1737
```

After running setup again, I noticed this was in zp-core/utilities/error_log-

`[15-Aug-2018 00:25:40 America/Chicago] PHP Fatal error:  Out of memory (allocated 4194304) (tried to allocate 8192 bytes) in /fullpath/zenphototest/zp-core/class-search.php on line 1541`

Switching to PHP 7.1.4 in 1.5, and enabling Imagick, same blank page crash, this time no PHP error log or utilities error log.

I then upgraded to the latest support build, and same thing, blank pages, no error logs.

In PHP 7.1.14 it uses the newer Imagick module 3.4.3, but still the older ImageMagick 6.7.6-8 2016-07-19 Q16. I can see that it is available before I try to enable it in Zenphoto-

![imagick version in php 7 1 4](https://user-images.githubusercontent.com/10157440/44135119-8eb46ac4-a01c-11e8-9983-771d14539878.png)

The only thing I can say I did different before all of this started was changing the PHP version before trying the upgrade to the minimum you guys support- 5.6, and after the install, I enabled https in Zenphoto.

**Comment by acrylian** [comment]
I fear I have no idea. It may be a configuration issue with your php version and imagick. Maybe something is not setup correctly. We didn't change any handling regarding the image processing actually (we did some things wrong with the cache file names but are working on getting those back correctly). 

We generally develop on PHP 7.2+ by now and 5.6 is the oldest we could test on if needed. But as mentioned on the 1.5 release post our code is by no means 5.6+ or 7+ only - we only rewrite things if applicable - but we simply have no resources to test on older versions. 

The only other thing that comes to mind is issues with the image itself. Although gifs normally have no exit or similar metadata it could be something with the image that is not right. That's why we ask for the images in question to test. If it works on our local and live setup it might be your server.

However look where your fatal error are reported:

- `class-album.php on line 1556`: That is within the Dynamic album class constructor and refers to database action. Do you heavily rely on dynamic albums and have a lot of images?
- `lib-auth.php on line 538`: That is within a method `getRights()` where I don't understand yet any memory issue could occur. It's just the definition method actually.
- `template-functions.php on line 1737` That is within the function `getAlbumPage()` which also is database related as it calculates - as the name says - the page number an image is on within an album: https://docs.zenphoto.org/function-getAlbumPage.html

Could you please perform a database refresh on the overview utilities. Perhaps your database is out of sync with the file system. That also can cause any sort of issues. Also clear the search cache if you have enabled it at all.

**Comment by MarkRH** [comment]
I tried this on my webhost which uses:

Linux
Apache Version | 2.4.34
PHP version: 5.6.37
Graphics support: PHP Imagick library 3.4.3
ImageMagick 6.7.2-7 2017-03-22 Q16 http://www.imagemagick.org	
    supporting: bmp, bmp2, bmp3, gif, gif87, jpeg, jpg, png, png24, png32, png8, tiff, tiff64	

Also tried on local PC which is using:

Windows 7 Pro
Apache Version | Apache/2.4.34 (Win32) PHP/7.2.8 OpenSSL/1.1.0h
PHP version: 7.2.8
Graphics support: PHP Imagick library 3.4.3
ImageMagick 7.0.7-11 Q16 x86 2017-11-23 http://www.imagemagick.org	
    supporting: bmp, bmp2, bmp3, gif, gif87, jpeg, jpg, png, png24, png32, png8, tiff, tiff64	

Thumbnail creation worked fine on both.  Downloaded some Spock GIF from Twitter to test with.

**Comment by MarkRH** [comment]
@acrylian I downloaded my Imagick from https://mlocati.github.io/articles/php-windows-imagick.html and followed their instructions how to get it working locally.  Although, I did have to edit the php_imagick.dll with a HEX editor to correct the version number.  https://github.com/mkoppanen/imagick/issues/248

I have 128 MB memory limit on both but just normal use doesn't consume more than 5 MB.  Anyway..
